### PR TITLE
[TE] Runner to generate SLA based metric data missing alerts

### PR DESCRIPTION
**Pull request by akshayrai**
Implements the runner that generates Data SLA anomalies. DATA_MISSING anomalies are created if the data is not available for the sla detection window within the configured SLA. 

Also inclues:
* Logic to merge SLA anomaly with existing and mark them as parent-child.
* Unit tests covering various DATA MISSING use-cases

### CODE REVIEW

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/datasla/DatasetSlaTaskRunner.java [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+   * We say the data is partial if we do not have all the data points in the sla detection window.
+   * Or more specifically if we have at least 1 data-point missing in the sla detection window.
+   *
+   * For example:
+   * Assume that the data is delayed and our current sla detection window is [1st Feb to 3rd Feb). During this scan,
+   * let's say data for 1st Feb arrives. Now, we have a situation where partial data is present. In other words, in the
+   * current sla detection window [1st to 3rd Feb) we have data for 1st Feb but data for 2nd Feb is missing. This is the
+   * partial data scenario.
+   */
+  private boolean isPartialData(long datasetLastRefreshTime, DetectionPipelineTaskInfo info, DatasetConfigDTO datasetConfig) {
```
Comment (xiaohui-sun) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/datasla/DatasetSlaTaskRunner.java:221]: Why we care about partial data? In our current setting the endTime of detection window is current time, and it is very common that we have partial data.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/datasla/DatasetSlaTaskRunner.java [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+
+  /**
+   * Validates if the data is delayed or not based on the user specified SLA configuration
+   */
+  private boolean hasMissedSLA(long datasetLastRefreshTime, Map<String, Object> slaProps, long slaDetectionEndTime) {
+    // fetch the user configured SLA, otherwise default 3_DAYS.
+    long delay = TimeGranularity.fromString(MapUtils.getString(slaProps,  "sla", DEFAULT_DATA_SLA))
+        .toPeriod().toStandardDuration().getMillis();
+
+    return (slaDetectionEndTime - datasetLastRefreshTime) >= delay;
```
Comment (xiaohui-sun) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/datasla/DatasetSlaTaskRunner.java:234]: We need to consider the expectedDelay in datasetconfig.
There are some datasets that have several days delay by default.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/datasla/DatasetSlaTaskRunner.java [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+            if (hasMissedSLA(datasetLastRefreshTime, slaProps, info.getEnd())) {
+              createDataSLAAnomaly(datasetLastRefreshTime + 1, info.getEnd(), info.getConfigId(), datasetConfig, slaProps);
+            }
+          }
+        }
+      } else if (isPartialData(datasetLastRefreshTime, info, datasetConfig)) {
+        // Optimize for the common case - the common case is that the data availability events are arriving
+        // correctly and we need not re-fetch the data to double check.
+        if (hasMissedSLA(datasetLastRefreshTime, slaProps, info.getEnd())) {
+          createDataSLAAnomaly(datasetLastRefreshTime + 1, info.getEnd(), info.getConfigId(), datasetConfig, slaProps);
```
Comment (xiaohui-sun) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/datasla/DatasetSlaTaskRunner.java:195]: What we have so complicated logic here? Let's find some time to discuss offline.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/datasla/DatasetSlaTaskRunner.java [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+    existingAnomalies.removeIf(MergedAnomalyResultBean::isChild);
+    MergedAnomalyResultDTO existingParentSLAAnomaly = existingAnomalies.get(0);
+
+    if (isDuplicateSLAAnomaly(existingParentSLAAnomaly, anomaly)) {
+      // Ensure anomalies are not duplicated. Ignore and return.
+      // Example: daily data with hourly cron should generate only 1 sla alert if data is missing
+      return;
+    }
+    existingParentSLAAnomaly.setChild(true);
+    anomaly.setChild(false);
```
Comment (xiaohui-sun) [inline thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/datasla/DatasetSlaTaskRunner.java:261]: Why this is called "merge"? It is only dedup?

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/datasla/DatasetSlaTaskRunner.java (hunk 1 above)
Comment (akshayrai) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/datasla/DatasetSlaTaskRunner.java:221]: discussed offline! others - pls refer to the comment in the code.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/datasla/DatasetSlaTaskRunner.java (hunk 3 above)
Comment (akshayrai) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/datasla/DatasetSlaTaskRunner.java:195]: discussed offline

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/datasla/DatasetSlaTaskRunner.java (hunk 4 above)
Comment (akshayrai) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/datasla/DatasetSlaTaskRunner.java:261]: Discussed offline. When 2 sla anomalies overlap (not duplicate) we merge them into a parent child relationship.

File: thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/datasla/DatasetSlaTaskRunner.java (hunk 2 above)
Comment (akshayrai) [reply thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/datasla/DatasetSlaTaskRunner.java:234]: The default dataset delay logic will apply when we automatically onboard all the alerts for SLA detection. The yaml translation logic needs to take care of it. 

Here, we expect that the sla is defined by the user and we will strictly adhere to it.

**Pull request merged**